tests/python/test_spark/xgboost_local_cluster_test.py

```diff
@@ -0,0 +1,465 @@
+
+    def test_regressor_basic_with_params(self):
+        regressor = XgboostRegressor(**self.reg_params)
+        model = regressor.fit(self.reg_df_train)
+        pred_result = model.transform(self.reg_df_test).collect()
+        for row in pred_result:
+            self.assertTrue(
+                np.isclose(
+                    row.prediction, row.expected_prediction_with_params, atol=1e-3
+                )
+            )
+
+    def test_callbacks(self):
+        from xgboost.callback import LearningRateScheduler
+
+        path = os.path.join(self.tempdir, str(uuid.uuid4()))
+
+        def custom_learning_rate(boosting_round):
+            return 1.0 / (boosting_round + 1)
+
+        cb = [LearningRateScheduler(custom_learning_rate)]
+        regressor = XgboostRegressor(callbacks=cb)
+
+        # Test the save/load of the estimator instead of the model, since
+        # the callbacks param only exists in the estimator but not in the model
+        regressor.save(path)
+        regressor = XgboostRegressor.load(path)
+
+        model = regressor.fit(self.reg_df_train)
+        pred_result = model.transform(self.reg_df_test).collect()
+        for row in pred_result:
+            self.assertTrue(
+                np.isclose(
+                    row.prediction, row.expected_prediction_with_callbacks, atol=1e-3
+                )
+            )
+
+    def test_classifier_distributed_basic(self):
+        classifier = XgboostClassifier(num_workers=self.n_workers, n_estimators=100)
+        model = classifier.fit(self.cls_df_train_distributed)
+        pred_result = model.transform(self.cls_df_test_distributed).collect()
+        for row in pred_result:
+            self.assertTrue(np.isclose(row.expected_label, row.prediction, atol=1e-3))
+            self.assertTrue(
+                np.allclose(row.expected_probability, row.probability, atol=1e-3)
+            )
+
+    def test_classifier_distributed_multiclass(self):
+        # There is no built-in multiclass option for external storage
+        classifier = XgboostClassifier(num_workers=self.n_workers, n_estimators=100)
+        model = classifier.fit(self.cls_df_train_distributed_multiclass)
+        pred_result = model.transform(self.cls_df_test_distributed_multiclass).collect()
+        for row in pred_result:
+            self.assertTrue(np.isclose(row.expected_label, row.prediction, atol=1e-3))
+            self.assertTrue(
+                np.allclose(row.expected_margins, row.rawPrediction, atol=1e-3)
+            )
+
+    def test_regressor_distributed_basic(self):
+        regressor = XgboostRegressor(num_workers=self.n_workers, n_estimators=100)
+        model = regressor.fit(self.reg_df_train_distributed)
+        pred_result = model.transform(self.reg_df_test_distributed).collect()
+        for row in pred_result:
+            self.assertTrue(np.isclose(row.expected_label, row.prediction, atol=1e-3))
+
+    @unittest.skip
+    def test_check_use_gpu_param(self):
```

Comment:
Q: shall I setup a CI test to run over GPU cluster ?

Reply:
Yes. Please place GPU related tests under tests/python-gpu/

Reply:
@trivialfis Where can I see the GPU CI machine config ? (e.g. how many GPU we can use on the CI machine)

Ideally, my GPU test I hope it can run on a machine with 4 GPUs (I will setup a spark cluster with local-cluster mode and run spark jobs with 4 executors, each spark executor will be allocated with 1 GPU.)

Reply:
@hcho3 May have more insight into the CI infrastructure.

> y GPU test I hope it can run on a machine with 4 GPUs

The mgpu instance should be g4dn.12xlarge on aws, which comes with 4 GPUs.

Reply:
@hcho3 May have more insight into the CI infrastructure.

> y GPU test I hope it can run on a machine with 4 GPUs

The mgpu instance should be g4dn.12xlarge on aws, which comes with 4 GPUs. You can mark the spark tests with `mgpu` pytest mark: https://github.com/dmlc/xgboost/blob/f470ad3af98e6632d81596c7241aa294b4a7648f/tests/python-gpu/conftest.py#L59 . https://github.com/dmlc/xgboost/blob/f470ad3af98e6632d81596c7241aa294b4a7648f/tests/pytest.ini#L3

Reply:
Thank you! @trivialfis 
An additional question, how to build xgboost python package with GPU support on my local machine ? `python setup.py bdist_wheel` is the command correct ?

Reply:
That command doesn't build with GPU support. If you want to build the wheel without installing XGBoost, then you need to do it in 2 steps: https://xgboost.readthedocs.io/en/stable/build.html#building-with-gpu-support .

``` sh
cd ./build
cmake -DUSE_CUDA=ON -DUSE_NCCL=ON /path/to/xgboost/
make -j$(nproc)
```

after which, you can create the binary wheel for Python:

```
cd python-package
python setup.py bdist_wheel
```

Reply:
Since you are writing Python code, I think the 2 step approach might be desired. You can compile C++ code once then run
``` sh
pip install . -e
```
and you don't have to run any build procedure anymore as long as you are testing locally.